### PR TITLE
Don't allow an empty assembly name to be stored in the workspace.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems?id=233669&_a=edit

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @jasonmalinowski @dotnet/roslyn-ide

**Comment by jmarolf** [comment]
>     public class ProjectInfoTests

Can we add a regression tests that verifies that empty names cannot be stored

---
Refers to: src/Workspaces/CoreTest/SolutionTests/ProjectInfoTests.cs:8 in 934f163. [](commit_id = 934f1634d2ca5dd72b48bce5df9af480b8a174e7, deletion_comment = False)

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski Please take a look.  

Here's my reasoning on why this was an ok change.  First, we always had similar behavior to this when using the MSBuildWorkspace.  Now, i'm just relaxing hte system to have this resiliency for all workspaces, not just hte MSBuildWorkspace.

**Comment by CyrusNajmabadi** [comment]
test this please

**Comment by CyrusNajmabadi** [comment]
test windows_debug_unit32_prtest  please

**Comment by CyrusNajmabadi** [comment]
ping @jasonmalinowski  @Pilchie

**Comment by jasonmalinowski** [comment]
I'm not sure I'm a fan of this given we're an API. Since passing null is indistinguishable from a legitimate programming bug I'd say we should fail fast with nulls (i.e. update the ProjectInfo constructor) and then expect workspaces to understand what they have to do. I'm sure if we did a poll and asked "what do you think `new ProjectInfo(assemblyName: null).AssemblyName` returns?" they sure wouldn't guess this.

I assume you've already done the analysis of how we got a null through here and that this doesn't mean we have a bug somewhere else in the project system, or there's some other rule we have to implement?

**Comment by CyrusNajmabadi** [comment]
Are you saying: FailFast on null, but do this on whitespace?  Or FailFast on all invalid assembly names?

I'm ok with doing that.  I'm just worried something else will crash higher up in the stack.  As this was late, i wasn't certain what to do?

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski ping when you have a moment.  I could use some fish teaching here.

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs [hunk 1]
```diff
@@ -380,7 +385,12 @@ internal class ProjectAttributes : IChecksummedObject, IObjectWritable
                 Id = id ?? throw new ArgumentNullException(nameof(id));
                 Name = name ?? throw new ArgumentNullException(nameof(name));
                 Language = language ?? throw new ArgumentNullException(nameof(language));
-                AssemblyName = assemblyName ?? throw new ArgumentNullException(nameof(assemblyName));
+
+                // Empty assembly names are not allowed.  For now just default to an unused assembly
+                // name.
+                AssemblyName = string.IsNullOrWhiteSpace(assemblyName)
+                    ? InvalidAssemblyNamePrefix + Interlocked.Increment(ref s_nextInvalidAssemblySuffix)
+                    : assemblyName;
```
Comment (tmeschter) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs]: I think we should throw an exception if `language`, or `assemblyName` are null or the empty string. Maybe even `name` though the implications of that aren't clear to me.

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs]: Well, that's the issue we're running into today.  We're throwing an exception, and it's causing problems :D

**Comment by CyrusNajmabadi** [comment]
@OmarTawfik Are you making it so that diagnostics can be stored inside Options?  If so, it seems like it would be great if 'empty assembly name' was just a diagnostic produced on the compiler side.  Can i just assign the VS bug over to you?

**Comment by OmarTawfik** [comment]
@CyrusNajmabadi the change i'm doing is to ParseOptions, not to CompilationOptions. The later already have diagnostics associated with them. So we can just add the check there (if it isn't existing already).
Sure, assign it to me :)

**Comment by CyrusNajmabadi** [comment]
great.  thanks!

**Comment by CyrusNajmabadi** [comment]
Oh.  Aleksey says he already did this :)  Just closing this out then!

**Pull request closed without merging**